### PR TITLE
Allow optype CREATE for append-only indexing operations

### PR DESCRIPTION
**Pull request by ywelsch**
Bulk requests currently do not allow adding "create" actions with auto-generated IDs. This PR allows using the optype CREATE for append-only indexing operations. This is mainly the user-facing aspect of it. Follow-ups will include:
- checking whether these should be handled differently at the engine level
- checking whether the default for `POST index/_doc` should be the optype CREATE.

### CODE REVIEW

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/2 
run elasticsearch-ci/bwc
run elasticsearch-ci/default-distro

**Comment by ywelsch** [comment]
@elasticmachine run elasticsearch-ci/default-distro

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 1]
```diff
@@ -3638,7 +3640,8 @@ public void testRetryWithAutogeneratedIdWorksAndNoDuplicateDocs() throws IOExcep
         long autoGeneratedIdTimestamp = 0;
 
         Engine.Index index = new Engine.Index(newUid(doc), doc, UNASSIGNED_SEQ_NO, 0,
-            Versions.MATCH_ANY, VersionType.INTERNAL, PRIMARY, System.nanoTime(), autoGeneratedIdTimestamp, isRetry, UNASSIGNED_SEQ_NO, 0);
+            Versions.MATCH_DELETED, VersionType.INTERNAL, PRIMARY, System.nanoTime(), autoGeneratedIdTimestamp, isRetry,
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I think both `MATCH_ANY` and `MATCH_DELETED` should work? If so, I suggest to pick one of them randomly.

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 2]
```diff
@@ -3677,7 +3680,8 @@ public void testRetryWithAutogeneratedIdsAndWrongOrderWorksAndNoDuplicateDocs()
         long autoGeneratedIdTimestamp = 0;
 
         Engine.Index firstIndexRequest = new Engine.Index(newUid(doc), doc, UNASSIGNED_SEQ_NO, 0,
-            Versions.MATCH_ANY, VersionType.INTERNAL, PRIMARY, System.nanoTime(), autoGeneratedIdTimestamp, isRetry, UNASSIGNED_SEQ_NO, 0);
+            Versions.MATCH_DELETED, VersionType.INTERNAL, PRIMARY, System.nanoTime(), autoGeneratedIdTimestamp, isRetry, UNASSIGNED_SEQ_NO,
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: I think both `MATCH_ANY` and `MATCH_DELETED` should work? If so, I suggest to pick one of them randomly.

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 3]
```diff
@@ -4784,7 +4790,7 @@ public void testSeqNoGenerator() throws IOException {
                     Versions.MATCH_ANY,
                     VersionType.INTERNAL,
                     Engine.Operation.Origin.PRIMARY,
-                    System.currentTimeMillis(), UNASSIGNED_SEQ_NO, 0);
+                    IndexRequest.UNSET_AUTO_GENERATED_TIMESTAMP, UNASSIGNED_SEQ_NO, 0);
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: One more `startTime`, I think this should be:

```suggestion
                    System.nanoTime(), UNASSIGNED_SEQ_NO, 0);
```

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 4]
```diff
@@ -4769,7 +4775,7 @@ public void testSeqNoGenerator() throws IOException {
                     VersionType.INTERNAL,
                     Engine.Operation.Origin.PRIMARY,
                     System.currentTimeMillis(),
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java]: Kind of unrelated, but given next comment, I think this `startTime` should also change into:

```suggestion
                     System.nanoTime(),
```

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/IndexingIT.java [hunk 5]
```diff
@@ -120,6 +125,60 @@ public void testIndexing() throws IOException {
...
+                createTestIndex.setJsonEntity("{\"settings\": {\"index.number_of_replicas\": 0}}");
+                client().performRequest(createTestIndex);
+                break;
+            case MIXED:
+                Request waitForGreen = new Request("GET", "/_cluster/health");
+                waitForGreen.addParameter("wait_for_nodes", "3");
+                client().performRequest(waitForGreen);
+
+                Version minNodeVersion = null;
+                Map<?, ?> response = entityAsMap(client().performRequest(new Request("GET", "_nodes/plugins")));
```
Comment (henningandersen) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/IndexingIT.java]: Unsure if `plugins` is a copy-paste error or if it has some benefit here?

**Review by henningandersen — COMMENTED** [review]
This is looking good already. I left a number of smaller comments to consider and/or address.

File: server/src/main/java/org/elasticsearch/action/index/IndexRequest.java [hunk 6]
```diff
@@ -215,8 +215,14 @@ public ActionRequestValidationException validate() {
             }
         }
 
-        if (opType() != OpType.INDEX && id == null) {
-            addValidationError("an id is required for a " + opType() + " operation", validationException);
+        if (id == null) {
+            if (opType != OpType.INDEX && opType != OpType.CREATE) {
+                addValidationError("an id is required for a " + opType + " operation", validationException);
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/index/IndexRequest.java]: turned this into an assertion.

File: server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java [hunk 7]
```diff
@@ -158,13 +158,18 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
...
         for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
             IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
             if (indexRequest != null) {
                 // Each index request needs to be evaluated, because this method also modifies the IndexRequest
                 boolean indexRequestHasPipeline = resolveRequiredOrDefaultPipeline(actionRequest, indexRequest, metaData);
                 hasIndexRequestsWithPipelines |= indexRequestHasPipeline;
             }
+
+            if (actionRequest instanceof IndexRequest) {
+                ((IndexRequest) actionRequest).checkAutoIdWithOpTypeCreateSupportedByVersion(minNodeVersion);
```
[the remark this answers is not in the document]
Comment (ywelsch) [reply server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java]: There's no strict policy on this. Given that it has the same HTTP code, and really about an API being wrongly used (something that you would not typically not react to in a scripted environment), I think this is fine.

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/IndexingIT.java (hunk 5 above)
Comment (ywelsch) [reply qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/IndexingIT.java]: copy paste error.

**Review by ywelsch — COMMENTED** [review]
Thanks for the thorough review. All addressed in 3a31040

**Review by henningandersen — APPROVED** [review]
LGTM.

I think we should hold off on backporting until we have ensured that the engine properly handles op_type create requests with auto-id.

**Pull request merged**